File: ts/package/src/info/handlers.ts
import * as rust from "../utils/rust-types.js";

import type { Flake, FlakeQuery, Group, User } from "./models.js";
import { SessionGet, type Sessioned } from "../utils/session.js";

export function scope() {
  return "/info";
}

export type flake_input = Sessioned<{ query: FlakeQuery }>;
export type flake_output = Flake;
export async function flake(input: flake_input): Promise<flake_output> {
  return SessionGet(input, scope(), "/flake");
}

export type users_input = Sessioned<{
  path: { scope: rust.String };
}>;
export type users_output = rust.Vec<User>;
export async function users(input: users_input): Promise<users_output> {
  return SessionGet(input, scope(), (path) => `/users/${path.scope}/users`);
}

export type groups_input = Sessioned<{
  path: { scope: rust.String };
}>;
export type groups_output = rust.Vec<Group>;
export async function groups(input: groups_input): Promise<groups_output> {
  return SessionGet(input, scope(), (path) => `/users/${path.scope}/groups`);
}

File: rust/package/src/info/handlers.rs
use crate::{
    info::{
        Group, User,
        models::{Flake, FlakeQuery},
    },
    utils::{Empty, SessionGetInput, SessionGetOutput, session_get},
};

pub fn scope() -> String {
    "/info".to_string()
}

pub type FlakeInput<'a> = SessionGetInput<'a, Empty, FlakeQuery>;
pub type FlakeOutput = Flake;
pub async fn flake(input: FlakeInput<'_>) -> SessionGetOutput<FlakeOutput> {
    session_get(input, scope(), |_path| "/flake").await
}

#[derive(Debug, Clone, PartialEq)]
pub struct UsersFilePath {
    pub scope: String,
}
pub type UsersInput<'a> = SessionGetInput<'a, UsersFilePath, Empty>;
pub type UsersOutput = Vec<User>;
pub async fn users(input: UsersInput<'_>) -> SessionGetOutput<UsersOutput> {
    session_get(input, scope(), |path| {
        format!("/users/{scope}/users", scope = path.scope)
    })
    .await
}

#[derive(Debug, Clone, PartialEq)]
pub struct GroupsFilePath {
    pub scope: String,
}
pub type GroupsInput<'a> = SessionGetInput<'a, GroupsFilePath, Empty>;
pub type GroupsOutput = Vec<Group>;
pub async fn groups(input: GroupsInput<'_>) -> SessionGetOutput<GroupsOutput> {
    session_get(input, scope(), |path| {
        format!("/users/{scope}/groups", scope = path.scope)
    })
    .await
}
